t/issue/1508.t

```diff
@@ -0,0 +1,165 @@
+
+=head1 NAME 
+
+issue/1508.t  - Check that the `needs()` function correctly propogates run-time task arguments
+
+=head1 DESCRIPTION
+
+Check that the `needs()` function does indeed correctly propogate 
+run-time task arguments (%params and @args) from the calling task down to the "needed" tasks.
+
+=head1 DETAILS
+
+  * AUTHOR / DATE : [tabulon]@[2021-09-26]
+  * RELATES-TO    : [github issue #1508](https://github.com/RexOps/Rex/issues/1508#issue-1007457392)
+  * INSPIRED from : t/needs.t
+
+=cut
+
```

Comment:
Hi @Ferki, 

Thank you for your through review -- of this PR as well as the related issue. 

I will try to briefly address your remarks in several chunks.

### POD in test files

Since the `.t` files are valid perl scripts, putting `POD` is technically allowed. 

For some kinds of content, I find it nicer than block/line comments. It renders better and folds neatly in the text editor. Whether or not it is socially allowed is up to the maintainer (you), so go ahead and yank it if you wish.

Reply:
##### Why were the tests marked as TODO ?

The cherry-picking thing might have played a role in there, yes. 
But that's also what I usually try to do, because I find it way cleaner way doing TDD. 

If `Rex` policy is as you describe, I will try to remember next time. It's true that this has become quite popular among many languages lately.

Meanwhile, here are my reasons, in a nutshell, in an attempt to convince you of the added value:

* Marking tests that are expected to fail (as such) is an extra piece of valuable information.

* Perl (with its excellent Test::* modules) had the ability to do TDD way before other languages. TODO tests were invented as part of this scheme, and they have quite interesting characteristics. 

* When you do a normal `prove`, TODO tests still get executed (and their results are briefly reported), but their failure won't cause the whole test suite to fail!

* When you do a `prove -v` (verbose mode), TODO tests behave as as normal tests : they will get detailed reporting 

* If TODO tests actually pass (instead of failing), that's also reported as a surprise (so that you can suspect something fishy)

* Also, the cherry-picking of solutions can then happen across PRs. 

Basically, it tends to be a superior form of TDD, in my humble opinion: 

- You can start with writing TODO tests for a given issue (without necessarily committing any code that satisfies them) ; keep them around, even ship them as such. 
- Months later, you (or someone else) can come up with one or more solutions which can easily be checked with `prove -v`. 
- Once satisfied with a solution, you unmark the TODO tests within the PR that commits the solution.

And, of course, all of that can happen within the same PR (as was the case here).

But you probably already know all this... And you probably have good reasons to prefer otherwise.

In the end, whatever the maintainer of the project decides, of course.

Reply:
# POD in tests

In general I'm not opposed to add POD to test files, provided we can realistically commit to establish it as the new norm, _and_ also to start adding them retroactively for consistency. I feel that would involve a considerable overhead, so I'd wish to discuss that topic outside of the context of this PR (and focus on passing arguments to `needs` here).

---

# TODO tests

Having new tests that first pass, which then start to fail when I successfully implement their logic feels backward _for my brain_ :upside_down_face: I can accept it works better for you, and I appreciate your detailed reasoning :heart: 

So far I used `TODO` for tests that are expected to fail, _and_ not intend to fix in the same PR that introduces them. So I'm fine shipping TODO tests (e.g. there's one in `t/file.t`), and I'd even prefer it over things like "skip this on Windows" whenever possible (the current test suite mostly SKIPs though :/ ).

In the usual PR context, I like to have it demonstrated by the CI run that the new tests are actually failing before the related change gets implemented. So I push the "Add tests for X" commits early, and watch them fail ("red"). Then I also like to demonstrate that the new commit (or sometimes series of commits) I push to implement/fix the failing behavior actually makes the test suite to pass ("green"). Then if I need to simplify the code, I can make that at the end, and demonstrate the tests still pass ("refactor")

If my PR would be about adding new tests that I don't intend to fix in the same PR, I think I'd still push a commit with the failing tests first, then make the CI pass with a second commit that marks them as TODO.

I'd be happy to further discuss these methodology details via [GitHub discussions](https://github.com/RexOps/Rex/discussions) or chat.